src/dune_pkg/toolchain.mli

```diff
@@ -0,0 +1,33 @@
+open! Import
+open! Stdune
+
+(* [Compiler.t] represents a reference to a compiler installation in a shared
+   location.
+
+   It can be retrieved using [get] at which point it will either be returned
+   if it is installed or downloaded and built.
+*)
+module Compiler : sig
+  type uninit
```

Comment:
How about just combining find and get so that you don't need this?

Reply:
Personally I like the fact that logic for looking up a compiler by name/version is separated from the logic for installing the compiler.

What do you see as the advantage of not exposing the uninstalled and installed types?

Reply:
When adding new code, IMHO we should justify why we're adding the code rather than looking for reasons not to add it. So I would like to invert your question:

* Why do we need both an installed and uninstalled type?
* Can we do anything useful to an uninstalled toolchain?

If there are no clear answers to the questions above, then I would suggest the interface is not as simple as it could be. I'm usually biased towards minimalism, so I'm willing to be flexible and entertain the idea that I'm missing the point. However, I would like to point an example where I managed to eliminate such a seemingly unnecessary type and greatly simplify and clarify the interface in the process: https://github.com/ocaml/dune/pull/10071 (See `Remote.uninit`). I'm not saying this situation is analogous, but I have my suspicions.

If you are convinced that this type is necessary, feel free to resolve this comment.

Reply:
Strictly speaking having a separate type for installed and uninstalled compilers is not necessary. However it allows us to separate the logic for choosing which compiler to install from the logic for installing a compiler, and I believe that this justifies the extra code. When consuming APIs I find that separating concerns like this helps me understand how to use them, and also helps me understand how to debug when things go wrong. The failure mode for choosing which compiler to install is very different from the failure mode for installing the compiler, and I expect it to aid debugging if these tasks are performed by different functions.

Reply:
Sure, feel free to keep the separate functions. However, when integrating two different components, the amount of implementation details visible to each component needs to be minimized. I think the current API does not achieve that goal. If this is because some other API quality was prioritized ahead of that, then I think it's regrettable.